### ISSUE TITLE
enable tbb

### ISSUE THREAD

**Issue by pca006132** [openscad/openscad-playground#59]
Note that manifold can now be compiled with tbb with minimal changes: https://github.com/elalish/manifold/pull/1045

This can probably enable openscad build with tbb as well. I can help if anyone is interested in adding it here.

**Comment by ochafik** [comment]
Hi @pca006132, thanks for the heads up!

We'll want need to update https://github.com/openscad/openscad-wasm with the oneTBB dep in our build image (cc/ @DSchroer, @t-paul), but in the meantime it seems to build fine:

- Remove `set(MANIFOLD_PAR OFF CACHE BOOL "" FORCE)` from OpenSCAD's CMakeLists.txt
- Use the helper from https://github.com/openscad/openscad/pull/5210 to build oneTBB (using [flags from their wiki](https://github.com/uxlfoundation/oneTBB/blob/master/WASM_Support.md)) and openscad wasm:

```bash
./scripts/wasm-base-docker-run.sh /bin/bash -c ' \
    ( test -d oneTBB || git clone https://github.com/uxlfoundation/oneTBB ) && \
    cd oneTBB && \
    rm -fR build && \
    emcmake cmake -B build \
        -DBUILD_SHARED_LIBS=ON \
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_CXX_FLAGS="-Wno-unused-command-line-argument" \
        -DCMAKE_EXE_LINKER_FLAGS="-sMALLOC=emmalloc" \
        -DTBB_DISABLE_HWLOC_AUTOMATIC_SEARCH=ON \
        -DTBB_EXAMPLES=OFF \
        -DTBB_STRICT=OFF \
        -DTBB_TEST=OFF && \
    ( cmake --build build -j || cmake --build build -j2 || cmake --build build ) && \
    cmake --install build && \
    cd .. &&
    emcmake cmake -B build -DCMAKE_BUILD_TYPE=Release -DEXPERIMENTAL=1 && \
    ( cmake --build build -j || cmake --build build -j2 || cmake --build build )'
```

Then in the playground codebase: `mkdir -p libs/openscad-wasm && cp ../path/to/openscad/build/openscad.* libs/openscad-wasm/`

Now based on [emscripten's docs](https://emscripten.org/docs/porting/pthreads.html#pthreads-support) I guess to get any kind of parallelism we either need to add COOP / COEP headers and build both tbb & openscad with extra `-DCMAKE_CXX_FLAGS="-pthread" -DCMAKE_EXE_LINKER_FLAGS="-sPTHREAD_POOL_SIZE=4"` flags, or with `-DCMAKE_EXE_LINKER_FLAGS="--proxy-to-worker"` when headers aren't an option.

It would be nice to test both options with heavy models that leverage parallelism well: suggestions welcome :-)

**Comment by pca006132** [comment]
Yeah, and we do need mimalloc for performance here (see [allocator-performance](https://emscripten.org/docs/porting/pthreads.html#allocator-performance) section).

Do note that we encountered some weird stack corruption that are basically impossible to debug due to the lack of sanitizer support in wasm, not sure if this is fixed in newer versions of emscripten.
